### DISCUSSION TITLE
How to get the current file being modified in ContextHook?

### DISCUSSION THREAD

**Discussion by macintacos** [copier-org/copier#695] [Q&A]
Hello!

I was looking into using [ContextHook](https://copier.readthedocs.io/en/latest/faq/#how-can-i-alter-the-context-before-rendering-the-project), and it works great, but I found that because it gets run every time `copier` is about to modify a file, if you add any network requests into this context hook things can get quite slow.

I was trying to see if I could limit the network requests so that it only happened when necessary (when editing a very specific file), but I can't seem to figure out how to get the current file being edited when the `ContextHook`'s `hook` method is invoked. Is that just not possible, or am I perhaps missing something?

**Comment by pawamoy** [comment]
The context hook is indeed executed before each file is rendered. We should add visible warning about this in its docs.

Another workaround for you maybe: you could try to cache the result of the request so that it is ran only once through the whole build?
You could even cache the result of the hook itself.

I'll try to see if there's a way to provide more metadata in the context hook 🙂

**Reply by macintacos** [reply]
Adding more metadata would be fantastic. I'm trying some ways of caching to get around this, will report back if I land on a solution that I think is ok!

**Reply by pawamoy** [reply]
Great! Could you open a feature request on the copier-templates-extensions repository in fact? This way I can assign myself so that I don't forget about it :slightly_smiling_face:

**Reply by macintacos** [reply]
I solved it by adding the call in the ContextUpdater's `__init__` method, and it worked like a charm. My code looks something like the following:

```python
class ContextUpdater(ContextHook):
    """Handles updating the context before the project gets generated."""

    def __init__(self, environment):  # noqa
        super().__init__(environment)
        # Call once, cache result for future invocations
        self.cached_result: str = self.make_very_expensive_call_and_cache_result()

    def hook(self, context: dict) -> dict:
        """This method is called just before the context is used to modify the project.

        Args:
            context (dict): The dictionary after all of the answers have been given by
                the user. Modify this as needed.
        """
        context["cached_result"] = self.cached_result 
```

I'll open an issue in that repository momentarily!

**Reply by macintacos** [reply]
Filed: https://github.com/copier-org/copier-templates-extensions/issues/2
